app/lib/shared/count_topics.dart

```diff
@@ -32,3 +44,112 @@ Future<void> countTopics() async {
   await uploadBytesWithRetry(
       reportsBucket, topicsJsonFileName, jsonUtf8Encoder.convert(topics));
 }
+
+typedef CanonicalTopic = ({
+  String topic,
+  String description,
+  Set<String> aliases,
+});
+
+final canonicalTopics = () {
+  try {
+    final f = File(p.join(resolveAppDir(), '../doc/topics.yaml'));
+    final root = loadYamlNode(f.readAsStringSync(), sourceUrl: f.uri);
+    if (root is! YamlMap) {
+      throw SourceSpanFormatException('expected a map', root.span);
+    }
+    if (root.keys.length > 1) {
+      throw SourceSpanFormatException(
+        'only the "topic" key is allowed',
+        root.span,
+      );
+    }
+
+    final topics = root.expectListOfObjects('topics');
+    return UnmodifiableListView(topics.map<CanonicalTopic>((entry) {
+      final topic = entry.expectString('topic', maxLength: 32);
+      if (!isValidTopicFormat(topic)) {
+        throw SourceSpanFormatException('must be valid topic', entry.span);
+      }
+      if (entry.keys.length > 3) {
+        throw SourceSpanFormatException(
+          'only keys "topic", "description" and "aliases" are allowed',
+          entry.span,
+        );
+      }
+      return (
+        topic: topic,
+        description: entry.expectString('description', maxLength: 160),
+        aliases: Set.unmodifiable(entry.expectList('aliases').nodes.map((node) {
+          final value = node.value;
+          if (value is! String) {
+            throw SourceSpanFormatException('must be a string', node.span);
+          }
+          if (!isValidTopicFormat(value)) {
+            throw SourceSpanFormatException('must be valid topic', node.span);
+          }
+          return value;
+        })),
+      );
+    }).toList(growable: false));
+  } on Exception catch (e, st) {
+    _log.shout('failed to load doc/topics.yaml', e, st);
+
+    // This is sad, but we can just ignore it!
+    return UnmodifiableListView(<CanonicalTopic>[]);
+  }
+}();
+
+/// True, if [topic] is formatted like a valid topic.
+bool isValidTopicFormat(String topic) =>
+    RegExp(r'^[a-z0-9-]{2,32}$').hasMatch(topic) &&
```

Comment:
I think it lives in pkg/pub_package_reader, I think this is so trivial I just duplicated it rather than add another dependency.

We can change it if we want to. This is really just to test the configuration file.

Reply:
It has a `Iterable<ArchiveIssue> checkTopics(String pubspecContent)` function, which IMO should be private (or in `src/`) because the only thing `pub_package_reader` should expose is `Future<PackageSummary> summarizePackageArchive(...)`.

So even if there was a good method we did export that we could use, I think it would be wrong to do so. `pub_package_reader` is for reading packages and validating them.

One could argue we should have a separate package or library somewhere called `topic_validation.dart` -- but we don't, and I'm not sure where it should live :D